### PR TITLE
For #14239: Notification for QR scan when permissions have been denied

### PR DESCRIPTION
**Pull request by eliserichards**
For #14239 

App permissions in settings: https://developer.android.com/reference/android/provider/Settings#ACTION_APPLICATION_DETAILS_SETTINGS 
Sumo article for versions lower than P (Android 9, API 28): https://support.mozilla.org/1/mobile/%VERSION%/%OS%/%LOCALE%/qr-camera-access

todo
- [x] Search fragment
- [x] Search dialog fragment
- [x] FxA pair `PairFragment`
- [ ] tests
- [x] dialog pops up as soon as you deny permissions. use `permissionDidUpdate`
- [ ] put duplicate dialog code - where should this live? utils?

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt
Comment (kglazko) [inline app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt:365]: Maybe for testing purposes it would be good to decouple creating and showing from the same function? And the bulk of the function does seem to be mainly about creating dialog, and it is only showed at the end. If not, some comments would definitely improve the readibility here.

File: app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt
Comment (kglazko) [inline app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt:343]: Maybe needs to be renamed, it isn't intuitive to me right away what this variable means (I assumed the result of whether the permissions have been granted?) Maybe something like permissionsResults or permissionsGrantedResults (kind of long)

File: app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt
Comment (kglazko) [inline app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt:394]: Question: what are the scenarios when this would be called and view would be null? Just wanted to make sure I understand what's going on here.

File: app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt
Comment (kglazko) [inline app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt:380]: Would benefit from some comments about what's going on here with the URI library

File: app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt
Comment (kglazko) [inline app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt:378]: Needs a comment about why these specific versions (>= P) are being acted on in the if-statement for the ease of future debugging

File: app/src/main/java/org/mozilla/fenix/settings/PairFragment.kt
Comment (kglazko) [inline app/src/main/java/org/mozilla/fenix/settings/PairFragment.kt:132]: Likewise here, why Version M?

File: app/src/main/java/org/mozilla/fenix/settings/PairFragment.kt
Comment (kglazko) [inline app/src/main/java/org/mozilla/fenix/settings/PairFragment.kt:33]: I noticed this variable is being set in multiple files, is it safe to keep track of its state across all these different files and initialize it to false each time?

**Review by kglazko — COMMENTED** [review]
Very good! Some tests could definitely benefit this PR if possible. I left a few nits about the same variable being created and set to false across multiple files, and some long functions that maybe could benefit from being broken up for more testability, or perhaps just more comments, and a few more nits, but overall great job!

File: app/src/main/java/org/mozilla/fenix/settings/PairFragment.kt
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/PairFragment.kt:132]: This specific intent, `Settings.ACTION_APPLICATION_DETAILS_SETTINGS`, was added in Android 6.0/Marshmallow/M. So for phones running older versions of android, we're just going to link to the help page in SUMO rather than supporting older intents

File: app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt:378]: Note to self: this should be M, not P

File: app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt:343]: unfortunately it's an `override` function, so I can't rename 😢 however, I can add some comments & kdocs

File: app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/search/SearchFragment.kt:394]: Great catch! This was a copy/paste from `onBackPressed` above. They both should be `toolbarView.view.search_scan_button` or just `search_scan_button`

File: app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt
Comment (kglazko) [inline app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt:52]: Remove the comments here

File: app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt
Comment (kglazko) [reply app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt:365]: No worries then :D

File: app/src/main/java/org/mozilla/fenix/settings/PairFragment.kt
Comment (kglazko) [reply app/src/main/java/org/mozilla/fenix/settings/PairFragment.kt:153]: Awesome, thank you for the comment

**Review by kglazko — COMMENTED** [review]
Looks great, just a few nits and then squash it 👍

**Comment by kglazko** [comment]
Needs a quick lint fix: [task 2020-09-10T22:08:24.665Z] /builds/worker/checkouts/src/app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt:7:1: Unused import
[task 2020-09-10T22:08:24.665Z] /builds/worker/checkouts/src/app/src/test/java/org/mozilla/fenix/search/DefaultSearchControllerTest.kt:8:1: Unused import

**Comment by kglazko** [comment]
Awesome, I will approve but wait on CI before merging to make sure there aren't any further fixes needed

**Review by kglazko — APPROVED** [review]

**Pull request merged**
